src/core/env.ml

```diff
@@ -1,79 +1,98 @@
+(** [add v d a env] extends the environment [env] by mapping the string
+    [Bindlib.name_of v] to [(v,d,a)]. *)
+let add : tvar -> bool -> tbox -> env -> env = fun v d a env ->
+  (Bindlib.name_of v, (v, d, a)) :: env
 
 (** [find n env] returns the Bindlib variable associated to the variable  name
     [n] in the environment [env]. If none is found, [Not_found] is raised. *)
 let find : string -> env -> tvar = fun n env ->
-  fst (List.assoc n env)
+  let (x,_,_) = (List.assoc n env) in x
 
-(** [to_prod env t] builds a sequence of products whose  domains  are the
-    variables of the environment [env] (from left to right), and which body is
-    the term [t]: [to_prod [(xn,an);..;(x1,a1)] t = ∀x1:a1,..,∀xn:an,t]. *)
+(** [to_prod env t] builds a sequence of products whose domains are the {e not
+    defined} variables of the environment [env] (from left to right), and
+    which body is the term [t]: [to_prod [(xn,an);..;(x1,a1)] t =
+    ∀x1:a1,..,∀xn:an,t]. *)
 let to_prod : env -> tbox -> term = fun env t ->
-  let fn t (_,(x,a)) = _Prod a (Bindlib.bind_var x t) in
+  let fn t (_,(x,d,a)) = if d then t else _Prod a (Bindlib.bind_var x t) in
   Bindlib.unbox (List.fold_left fn t env)
 
-(** [to_abst env t] builds a sequence of abstractions whose  domains  are the
-    variables of the environment [env] (from left to right), and which body is
-    the term [t]: [to_prod [(xn,an);..;(x1,a1)] t = λx1:a1,..,λxn:an,t]. *)
+(** [to_abst env t] builds a sequence of abstractions whose domains are the {b
+    not defined} variables of the environment [env] (from left to right), and
+    which body is the term [t]: [to_prod [(xn,an);..;(x1,a1)] t =
+    λx1:a1,..,λxn:an,t]. *)
 let to_abst : env -> tbox -> term = fun env t ->
-  let fn t (_,(x,a)) = _Abst a (Bindlib.bind_var x t) in
+  let fn t (_,(x,d,a)) = if d then t else _Abst a (Bindlib.bind_var x t) in
   Bindlib.unbox (List.fold_left fn t env)
 
 (** [vars env] extracts the array of the Bindlib variables in [env]. Note that
     the order is reversed: [vars [(xn,an);..;(x1,a1)] = [|x1;..;xn|]]. *)
 let vars : env -> tvar array = fun env ->
-  Array.of_list (List.rev_map (fun (_,(x,_)) -> x) env)
+  let f (_, (x, d, _)) = if d then None else Some(x) in
+  Array.of_list (List.filter_rev_map f env)
 
-(** [to_term env] extracts the array of the variables in [env] and inject them
-    in the [tbox] type. This is the same as [Array.map _Vari (vars env)]. Note
-    that the order is reversed: [vars [(xn,an);..;(x1,a1)] = [|x1;..;xn|]]. *)
+(** [to_term env] extracts the array of the {e not defined} variables in [env]
+    and injects them in the [tbox] type.  This is the same as [Array.map _Vari
+    (vars env)]. Note that the order is reversed: [vars [(xn,an);..;(x1,a1)] =
+    [|x1;..;xn|]]. *)
 let to_tbox : env -> tbox array = fun env ->
-  Array.of_list (List.rev_map (fun (_,(x,_)) -> _Vari x) env)
+  let f (_, (x, d, _)) = if d then None else Some(_Vari x) in
```

Comment:
to be fixed as well

Reply:
sorry, not this one

Reply:
Really? what should we do when a variable is defined?

Reply:
There is nothing to change in this function I think. It just gathers an array with all the undefined variables. Do you think that we should do something else?

Reply:
No, there was a mistake of synchronisation of messages (I didn't see the
> sorry, not this one

)